### PR TITLE
1316: support resilient redis mode for cache util

### LINKED ISSUE #1316: Threescale Util fails the current request if redis connection is not initiated successfully
[Redis utility here](https://github.com/3scale/APIcast/blob/master/gateway/src/apicast/threescale_utils.lua#L153-L156) seems to not be designed with resiliency in mind, in case redis fails to connect for any reason, subsequent call to `_M.error()` will invoke `ngx.say` and `ngx.exit` and terminate the current request.

##### Version

V2.9

##### Steps To Reproduce

1. Import `redis` from `threescale_utils` module.
2. call `connect_redis`, e.g. local red = ts:connect_redis()`.
3. run APICast instance not connected to redis (like staging) or simulate redis disconnect.
4. Entire request fails.

##### Current Result
Entire request terminates.

##### Expected Result
Provide at least an key of `option` argument to be like `secondary` or `resilient`, and if passed `_M.error()` is not invoke if redis is not available and request gracefully continues.

### PR DESCRIPTION
**Pull request by abarrak**
This resolves #1316 and make it useful to use the cache without fail in case of network disconnect, etc.

Tests passing:
_(results included in commit message)_
Screen Shot 2022-02-10 at 12 07 17 PM

### CODE REVIEW

**Review by unleashed — DISMISSED** [review]

**Comment by abarrak** [comment]
Thanks @eloycoto and @unleashed ..
Just covered both scenarios and made sure to keep backward compatibility as the default.

Tests run is passing:
```shell
$ make busted-util
EXTRA_CFLAGS="-DHAVE_EVP_KDF_CTX=1" /usr/local/openresty/luajit/bin/rover install --roverfile=gateway/Roverfile > /dev/null
/usr/local/openresty/luajit/bin/rover exec bin/busted "spec/threescale_utils_spec.lua" 
●●●
3 successes / 0 failures / 0 errors / 0 pending : 0.022514 seconds
```

**Comment by abarrak** [comment]
Guys I thought again about current implementation and think it's better to keep `_M.error()` signature intact and introduce a new function (e.g. `_M.silent_error()`) using same signature `...` and without nginx exit call.

This way will make the redis connect use either based on the `resilient` flag and without impacting original function usage.
Shared chunk in the function will be reused without duplication.

What do you think @unleashed @eloycoto @kevprice83 ?

File: gateway/src/apicast/threescale_utils.lua
Comment (kevprice83) [inline gateway/src/apicast/threescale_utils.lua:155]: @abarrak I think it's better to add this as another condition rather than a nested if statement. `if not ok and if not resilient then` because we only want to execute something when both return false or both return true. It's not super important and maybe @eloycoto might be able to comment whether the nested if statement could introduce any additional unwanted processing.

File: gateway/src/apicast/threescale_utils.lua
Comment (kevprice83) [inline gateway/src/apicast/threescale_utils.lua:203]: I would maybe rename this to `error_gracefully()` because it's not a silent error as we are going to log the failed connection to redis.

Comment (abarrak) [reply gateway/src/apicast/threescale_utils.lua:203]: Noted 📝

File: gateway/src/apicast/threescale_utils.lua
Comment (abarrak) [reply gateway/src/apicast/threescale_utils.lua:155]: And what would you suggest for the other case?
Does `elseif not ok then …` sound good enough?

Comment (kevprice83) [reply gateway/src/apicast/threescale_utils.lua:155]: I'd just suggest something like:

```
if not ok and not resilient then
  return nil, _M.error("failed to connect to redis on ", host, ":", port, ": ", err)
else
  return nil, _M.error_silently("failed to connect to redis on ", host, ":", port, ": ", err)
end
```

This might just be about style though so let¡s wait for a comment from someone more knowledgeable in Lua than myself because if it doesn't introduce unnecessary processing then we can leave it as you have already written it. I assume this is executed on every single request though right? That's why I was a little concerned, I don't see us opening  a connection with Redis and reusing that in this code, we open a new connection for every inbound request.

Comment (abarrak) [reply gateway/src/apicast/threescale_utils.lua:155]: Alright will wait for @eloycoto feedback.
I considered it in the first, but changed to the current structure, due the fact that original condition was just `if not ok` only, and using `else` alone here would catch the cases where response is `ok`.

Many thanks for the review comments so far 🌹 will make sure to implement it in upcoming commit.

**Comment by eguzki** [comment]
@kevprice83 please review this PR and we decide to merge or close it

File: gateway/src/apicast/threescale_utils.lua
Comment (abarrak) [reply gateway/src/apicast/threescale_utils.lua:127]: It should be set by calling sites to this utility as part of `option` table argument.
I pass it in a custom policy that rely on redis for some caching requirement.
This, defaults to null when it's nothing is set, hence preserving main usage and backward compatibility.

Comment (kevprice83) [reply gateway/src/apicast/threescale_utils.lua:127]: @abarrak could you document an example of how to pass this option? I think that would be really helpful for others who would also benefit from this feature.

**Comment by abarrak** [comment]
Hi @eguzki @kevprice83, could you decide on this PR clousre?
It was originally approved but got new changes after review.

**Pull request closed without merging**
